### PR TITLE
[SYCL] Implement atomic_memory_scope_capabilities device query for OpenCL and Level Zero

### PR DESCRIPTION
**Pull request by maarquitos14**
Adds support to query devices for `atomic_memory_scope_capabilities`. The backends supported are OpenCL and Level Zero. For the rest of backends, it has been left unsupported.

### CODE REVIEW

File: sycl/include/sycl/detail/pi.h [hunk 1]
```diff
@@ -561,6 +561,19 @@ constexpr pi_memory_scope_capabilities PI_MEMORY_SCOPE_WORK_GROUP = 0x04;
 constexpr pi_memory_scope_capabilities PI_MEMORY_SCOPE_DEVICE = 0x08;
 constexpr pi_memory_scope_capabilities PI_MEMORY_SCOPE_SYSTEM = 0x10;
 
+// CL equivalents are only available for OpenCL version 3.0
+#define PI_DEVICE_ATOMIC_MEMORY_CAPABILITIES 0x1063
```
Comment (AlexeySachkov) [inline sycl/include/sycl/detail/pi.h:565]: Why do we need this macro here? The only use I see in OpenCL PI plugin. I suggest to define it there directly or even better, use a macro from `CL/cl.h` header

File: sycl/plugins/opencl/pi_opencl.cpp [hunk 2]
```diff
@@ -283,8 +283,56 @@ pi_result piDeviceGetInfo(pi_device device, pi_device_info paramName,
...
   case PI_DEVICE_INFO_UUID:
   case PI_DEVICE_INFO_ATOMIC_MEMORY_ORDER_CAPABILITIES:
-  case PI_DEVICE_INFO_ATOMIC_MEMORY_SCOPE_CAPABILITIES:
     return PI_ERROR_INVALID_VALUE;
+  case PI_DEVICE_INFO_ATOMIC_MEMORY_SCOPE_CAPABILITIES: {
+    // Initialize result to minimum mandated capabilities according to
+    // SYCL2020 4.6.3.2
+    pi_memory_scope_capabilities result = PI_MEMORY_SCOPE_WORK_ITEM |
+                                          PI_MEMORY_SCOPE_SUB_GROUP |
+                                          PI_MEMORY_SCOPE_WORK_GROUP;
```
Comment (AlexeySachkov) [inline sycl/plugins/opencl/pi_opencl.cpp:292]: From [Table 21. Context information descriptors](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.context.info):

> `info::context::atomic_memory_scope_capabilities`
> Returns the set of memory scopes supported by atomic operations on all devices in the context, which is guaranteed to include `work_group`.

Why our default return value provides more guarantees than described in a SYCL spec?

File: sycl/plugins/opencl/pi_opencl.cpp [hunk 3]
```diff
@@ -283,8 +283,56 @@ pi_result piDeviceGetInfo(pi_device device, pi_device_info paramName,
...
+
+    pi_device_atomic_capabilities devCapabilities = 0;
+    if (devVer >= OCLV::V3_0) {
+      ret_err = clGetDeviceInfo(deviceID, PI_DEVICE_ATOMIC_MEMORY_CAPABILITIES,
+                                sizeof(pi_device_atomic_capabilities),
+                                &devCapabilities, nullptr);
+      if (ret_err != CL_SUCCESS)
+        return static_cast<pi_result>(ret_err);
+      assert(devCapabilities && PI_DEVICE_ATOMIC_SCOPE_WORK_GROUP &&
+             "Violates minimum mandated guarantee");
```
Comment (AlexeySachkov) [inline sycl/plugins/opencl/pi_opencl.cpp:309]:
```suggestion
      assert((devCapabilities & PI_DEVICE_ATOMIC_SCOPE_WORK_GROUP) &&
             "Violates minimum mandated guarantee");
```

`PI_DEVICE_ATOMIC_SCOPE_WORK_GROUP` is always `true`, I guess

File: sycl/plugins/opencl/pi_opencl.cpp [hunk 4]
```diff
@@ -283,8 +283,56 @@ pi_result piDeviceGetInfo(pi_device device, pi_device_info paramName,
...
+
+      // Because scopes are hierarchical, wider scopes support all narrower
+      // scopes. SUB_GROUP and WORK_ITEM were already included in the
+      // initialization, since WORK_GROUP is mandated minimum capality.
+      if (devCapabilities && PI_DEVICE_ATOMIC_SCOPE_DEVICE) {
+        result |= PI_MEMORY_SCOPE_DEVICE;
+      }
```
Comment (AlexeySachkov) [inline sycl/plugins/opencl/pi_opencl.cpp:316]: Oh, I suppose that's the answer to my question above about default value. Do you have a pointer to spec section which describes the hierarchical nature of those scopes?

File: sycl/plugins/opencl/pi_opencl.cpp [hunk 5]
```diff
@@ -283,8 +283,56 @@ pi_result piDeviceGetInfo(pi_device device, pi_device_info paramName,
...
+      // Just return minimum mandated capabilities for older versions.
+      // OpenCL 1.x minimum mandated capabilities are WORK_GROUP, we
+      // already initialized using it.
+      if (devVer >= OCLV::V2_0) {
+        // OpenCL 2.x minimum mandated capabilities are WORK_GROUP | DEVICE |
+        // ALL_DEVICES
+        result |= PI_MEMORY_SCOPE_DEVICE | PI_MEMORY_SCOPE_SYSTEM;
+      }
+    }
+    std::memcpy(paramValue, &result, sizeof(result));
```
Comment (AlexeySachkov) [inline sycl/plugins/opencl/pi_opencl.cpp:333]: Please note that `paramValue` may be `nullptr`. This API has number of out-parameters and we need to handle them properly:
- we need to set `paramValueSizeRet` if it is not `nullptr`
- we need to return `PI_INVALID_VALUE` if size in bytes specified by `param_value_size` is `< size` of return type as specified in the [Device Queries](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#device-queries-table) table and `param_value` is not a `NULL` value

File: sycl/plugins/opencl/pi_opencl.cpp (hunk 3 above)
Comment (maarquitos14) [reply sycl/plugins/opencl/pi_opencl.cpp:309]: This is to check that `PI_DEVICE_ATOMIC_SCOPE_WORK_GROUP` is enabled in `devCapabilities`. It has to be enabled, because it is minimum mandated capability.

File: sycl/plugins/opencl/pi_opencl.cpp (hunk 4 above)
Comment (maarquitos14) [reply sycl/plugins/opencl/pi_opencl.cpp:316]: This comes from OpenCL Spec: https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_ATOMIC_MEMORY_CAPABILITIES

However, you can see there that `WORK_ITEM` is actually an exception in OpenCL. I had an offline discussion with @gmlueck, @bashbaug and @Pennycook about this, and the conclusion is that SYCL should always return `WORK_ITEM`. I think there will be an update in the spec to detail this.

File: sycl/include/sycl/detail/pi.h (hunk 1 above)
Comment (andylshort) [reply sycl/include/sycl/detail/pi.h:565]: Could you not just include `CL/cl_version.h` and use `#ifdef CL_VERSION_3_0` (or similar) to surround where you use those symbols in the `if (devVer >= OCLV::V3_0) {` conditional block for `pi_opencl.cpp`? Would save having to add these redefinitions.

Comment (AlexeySachkov) [reply sycl/include/sycl/detail/pi.h:565]: > Could you not just use #ifdef CL_VERSION_3_0 (or similar) to surround where you use those symbols in the if (devVer >= OCLV::V3_0) { conditional block for pi_opencl.cpp?

Using such `#ifdef` would make sense if we were aiming to allow building our OpenCL plugin against different versions of OpenCL headers.

If we want to use 3.0 API, then we should probably update the following piece of code:

https://github.com/intel/llvm/blob/ccf205e4a281d47de4196a6b8aec2f9dc0277670/sycl/include/sycl/detail/cl.h#L11-L15

I honestly don't know if 3.0 headers are backwards compatible with previous OpenCL versions so I'm not entirely sure if anything would break.

The bare minimum which I would suggest is to at least move our custom define a level lower, because there seem to be no need to share it across all plugins or users of PI interface.

File: sycl/plugins/opencl/pi_opencl.cpp (hunk 4 above)
Comment (AlexeySachkov) [reply sycl/plugins/opencl/pi_opencl.cpp:316]: Got it, thanks! Perhaps it would be better to put that link to OCL spec directly into sources next to the comment

File: sycl/include/sycl/detail/pi.h (hunk 1 above)
Comment (bashbaug) [reply sycl/include/sycl/detail/pi.h:565]: > I honestly don't know if 3.0 headers are backwards compatible with previous OpenCL versions so I'm not entirely sure if anything would break.

The headers are fully backwards compatible.  There is actually only one set of OpenCL headers and the OpenCL version they support is controlled by an preprocessor define, see: https://github.com/KhronosGroup/OpenCL-Headers/#compiling-for-a-specific-opencl-version

File: sycl/plugins/opencl/pi_opencl.cpp (hunk 4 above)
Comment (maarquitos14) [reply sycl/plugins/opencl/pi_opencl.cpp:316]: I thought about that, but the OCL spec explicitly says `WORK_ITEM` is an exception, so that might be confusing because we are not treating it as an exception. When SYCL spec is updated, we can obviously add a ref to that next to the comment. For now, I'm not sure if referencing OCL spec is helpful or more confusing. What do you think?

File: opencl/CMakeLists.txt [hunk 6]
```diff
@@ -44,7 +44,7 @@ FetchContent_GetProperties(ocl-headers)
 set(OpenCL_INCLUDE_DIR
   ${ocl-headers_SOURCE_DIR} CACHE PATH "Path to OpenCL Headers")
 
-target_compile_definitions(Headers INTERFACE -DCL_TARGET_OPENCL_VERSION=220)
+target_compile_definitions(Headers INTERFACE -DCL_TARGET_OPENCL_VERSION=300)
```
Comment (AlexeySachkov) [inline opencl/CMakeLists.txt:47]: I don't think that this change is necessary. We can still built `opencl-aot` using 2.2 version of headers. Technically, the tool is only likely using 1.2 API

File: sycl/include/sycl/detail/cl.h [hunk 7]
```diff
@@ -11,7 +11,7 @@
 // Suppress a compiler message about undefined CL_TARGET_OPENCL_VERSION
 // and define all symbols up to OpenCL 2.2
```
Comment (AlexeySachkov) [inline sycl/include/sycl/detail/cl.h:12]:
```suggestion
// and define all symbols up to OpenCL 3.0
```

File: sycl/plugins/opencl/pi_opencl.cpp [hunk 8]
```diff
@@ -283,8 +283,63 @@ pi_result piDeviceGetInfo(pi_device device, pi_device_info paramName,
...
+      }
+    }
+    if (paramValue) {
+      if (paramValueSize < sizeof(cl_device_atomic_capabilities))
+        return PI_ERROR_INVALID_VALUE;
+
+      std::memcpy(paramValue, &result, sizeof(result));
+      if (paramValueSizeRet)
+        *paramValueSizeRet = sizeof(result);
+    }
```
Comment (AlexeySachkov) [inline sycl/plugins/opencl/pi_opencl.cpp:340]:
```suggestion
    }
    if (paramValueSizeRet)
      *paramValueSizeRet = sizeof(result);

```

`paramValueSizeRet` should be filled even if `paramValue` is nullptr.
That is needed to support (usually) initial query:
```
piDeviceInfoGet(device, paramName, nullptr, 0, &size_ret);
// now allocate size_ret bytes and query result
```

File: sycl/plugins/opencl/pi_opencl.cpp [hunk 9]
```diff
@@ -1787,7 +1842,7 @@ pi_result piextKernelGetNativeHandle(pi_kernel kernel,
 // Windows: dynamically loaded plugins might have been unloaded already
 // when this is called. Sycl RT holds onto the PI plugin so it can be
 // called safely. But this is not transitive. If the PI plugin in turn
-// dynamically loaded a different DLL, that may have been unloaded. 
+// dynamically loaded a different DLL, that may have been unloaded.
```
Comment (AlexeySachkov) [inline sycl/plugins/opencl/pi_opencl.cpp:1845]: Seems like an unrelated change, better to avoid

File: sycl/plugins/opencl/pi_opencl.cpp (hunk 4 above)
Comment (AlexeySachkov) [reply sycl/plugins/opencl/pi_opencl.cpp]: Technically, this code is within OpenCL plugin and therefore I think it shouldn't be too confusing to leave links to OpenCL spec in there.

**Review by AlexeySachkov — APPROVED** [review]
Overall LGTM

File: sycl/plugins/opencl/pi_opencl.cpp (hunk 4 above)
Comment (AlexeySachkov) [reply sycl/plugins/opencl/pi_opencl.cpp]: Discussed this offline. There is a corner case with `WORK_ITEM` scope being defined differently between SYCL and OpenCL, which is a reason for ignoring this OpenCL exception.

I think that it would still make sense to leave an OCL spec link, but we should have a comment about why we ignore that `work-item` exception. If we already have a link to SYCL spec clarification PR/issue, then even better, we can link it there as well.

**Comment by AlexeySachkov** [comment]
@dpcpp-esimd-reviewers, could you please take a look at `esimd_emulator` part of the PR? It is fairly simple

**Review by sarnex — APPROVED** [review]
esimd emu lgtm

File: opencl/CMakeLists.txt (hunk 6 above)
Comment (maarquitos14) [reply opencl/CMakeLists.txt:47]: We discussed this offline: we need to keep the change so that runtime itself is able to see definitions for OCL 3.0.

File: sycl/plugins/opencl/pi_opencl.cpp (hunk 4 above)
Comment (maarquitos14) [reply sycl/plugins/opencl/pi_opencl.cpp]: I extended the current comment with further explanation and a link to OCL spec. Please, let me know if this is enough.

**Comment by maarquitos14** [comment]
Friendly ping @intel/llvm-reviewers-runtime @intel/dpcpp-l0-pi-reviewers

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.cpp [hunk 10]
```diff
@@ -1164,9 +1166,17 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceGetInfo(
...
+  case UR_DEVICE_INFO_ATOMIC_MEMORY_SCOPE_CAPABILITIES: {
+    // There are no explicit restrictions in L0 programming guide, so assume all
+    // are supported
+    pi_memory_scope_capabilities result =
+        PI_MEMORY_SCOPE_WORK_ITEM | PI_MEMORY_SCOPE_SUB_GROUP |
+        PI_MEMORY_SCOPE_WORK_GROUP | PI_MEMORY_SCOPE_DEVICE |
+        PI_MEMORY_SCOPE_SYSTEM;
+
+    return ReturnValue(result);
+  }
```
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.cpp:1178]: @jandres742 : please review this

**Review by smaslov-intel — APPROVED** [review]
LGTM

File: sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.cpp (hunk 10 above)
Comment (jandres742) [reply sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero.cpp:1178]: assumption is correct for L0 GPU driver, based on what we return already in OpenCL GPU driver.

**Review by jandres742 — COMMENTED** [review]
+1

**Comment by maarquitos14** [comment]
@intel/llvm-reviewers-runtime assigned reviewer is on holiday, could you please take a look?

**Review by steffenlarsen — APPROVED** [review]
LGTM! 🚀

**Comment by bader** [comment]
@maarquitos14, `sycl/unittests/SYCL2020/AtomicMemoryScopeCapabilities.cpp` fails on Windows. Please, fix ASAP.
